2. For parts (a) through (c), indicate which of i. through iv. is correct.
Justify your answer.

(a) The lasso, relative to least squares, is:

i. More fexible and hence will give improved prediction accuracy when its increase in bias is less than its decrease in variance.

ANS: False

ii. More fexible and hence will give improved prediction accuracy when its increase in variance is less than its decrease in bias.

ANS: False

iii. Less fexible and hence will give improved prediction accuracy when its increase in bias is less than its decrease in variance.

ANS: TRUE

iv. Less fexible and hence will give improved prediction accuracy when its increase in variance is less than its decrease in bias.

ANS: False, as lasso will decrease the variance and increase the bias.

(b) Repeat (a) for ridge regression relative to least squares.

Less flexible and hence will give improved prediction accuracy when its increase in bias is less than its decrease in variance.

(c) Repeat (a) for non-linear methods relative to least squares.

More flexible and hence will give improved prediction accuracy when its increase in variance is less than its decrease in bias.

9. In this exercise, we will predict the number of applications received
using the other variables in the College data set.
(a) Split the data set into a training set and a test set.

In [33]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.linear_model import RidgeCV
from sklearn.linear_model import LassoCV

In [21]:
college_data = pd.read_csv('C:/Users/nihar/OneDrive/Desktop/Predictive Modeling/Data Sets/College.csv')

college_data = college_data.rename(columns={'Unnamed: 0': 'Name'})
college_data.head()

Name Private  Apps  Accept  Enroll  Top10perc  \
0  Abilene Christian University     Yes  1660    1232     721         23   
1            Adelphi University     Yes  2186    1924     512         16   
2                Adrian College     Yes  1428    1097     336         22   
3           Agnes Scott College     Yes   417     349     137         60   
4     Alaska Pacific University     Yes   193     146      55         16   

   Top25perc  F.Undergrad  P.Undergrad  Outstate  Room.Board  Books  Personal  \
0         52         2885          537      7440        3300    450      2200   
1         29         2683         1227     12280        6450    750      1500   
2         50         1036           99     11250        3750    400      1165   
3         89          510           63     12960        5450    450       875   
4         44          249          869      7560        4120    800      1500   

   PhD  Terminal  S.F.Ratio  perc.alumni  Expend  Grad.Rate  
0   70        78       18.1           12    7041         60  
1   29        30       12.2           16   10527         56  
2   53        66       12.9           30    8735         54  
3   92        97        7.7           37   19016         59  
4   76        72       11.9            2   10922         15

In [30]:
# Split the dataset into features (X) and target variable (y)
X = college_data.drop(['Apps','Name', 'Private'], axis=1)
y = college_data['Apps']

# Split the dataset into training set (80%) and test set (20%)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)



(b) Fit a linear model using least squares on the training set, and
report the test error obtained.

In [31]:

# Fit a linear model using least squares on the training set
model = LinearRegression()
model.fit(X_train, y_train)

# Predict on the test set
y_pred = model.predict(X_test)

# Calculate the mean squared error (test error)

mse = mean_squared_error(y_test, y_pred)

print(f"Test Error (MSE): {mse:.2f}")

Test Error (MSE): 1442624.91


(c) Fit a ridge regression model on the training set, with λ chosen by cross-validation. Report the test error obtained.


In [34]:
# Fit a ridge regression model on the training set with λ chosen by cross-validation
ridge_model = RidgeCV()
ridge_model.fit(X_train, y_train)

# Predict on the test set
y_pred_ridge = ridge_model.predict(X_test)

# Calculate the mean squared error (test error)
ridge_test_error = mean_squared_error(y_test, y_pred_ridge)
ridge_test_error

1442636.6703560795

(d) Fit a lasso model on the training set, with λ chosen by crossvalidation. Report the test error obtained, along with the number of non-zero coefcient estimates

In [35]:
# Fit a lasso model on the training set with λ chosen by cross-validation
lasso_model = LassoCV()
lasso_model.fit(X_train, y_train)

# Predict on the test set
y_pred_lasso = lasso_model.predict(X_test)

# Calculate the mean squared error (test error)
lasso_test_error = mean_squared_error(y_test, y_pred_lasso)
lasso_test_error

1587020.0176529177

In [36]:
# Number of non-zero coefficient estimates
num_non_zero_coef = len(lasso_model.coef_.nonzero()[0])
num_non_zero_coef

7

(g) Comment on the results obtained. How accurately can we predict the number of college applications received? Is there much diference among the test errors resulting from these five approaches?

Least Square Test R-Square: 1442624.91

Ridge Model Test R-Square: 1442636.6703560795

Lasso Model Test R-Square: 1587020.0176529177

It can be seen that the Lasso model predicts the highest R-Square. Though all the models have similar performance

This was expected as the minimum MSE was found for Lasso model across all the models